### PR TITLE
CLOUD-695: Fan-Out workspace triggers

### PR DESCRIPTION
**Pull request by chpatinos**
### Description
TFBuddy plan and apply events failed when an MR/PR touched many Terraform workspaces. `TriggerTFCEvents` processed workspaces sequentially inside the JetStream subscriber, so a batch of 5+ workspaces could exceed the default 30s `AckWait`. JetStream then redelivered the message and the duplicate workers raced on the same MR, leaving runs half-created with incomplete comments.

This MR replaces the synchronous loop with a fire-and-forget fan-out. The previous attempt at fixing this (#57) capped concurrency with an `errgroup` but kept everything inline, which only narrowed the timeout window without eliminating the AckWait class of failure.

### Testing

- `go test -race ./pkg/tfc_trigger/... ./pkg/tfc_api/...` — clean.
- End-to-end validated against a kind cluster with live NATS: `tilt up` brings up tfbuddy + NATS, the new stream and `tfbuddy_workspace_trigger_worker` consumer are registered on first boot, and a separate integration smoke test confirmed publish/subscribe + dedup behave correctly across "replicas" of the queue.

### CODE REVIEW

File: pkg/tfc_trigger/tfc_trigger.go [hunk 1]
```diff
@@ -373,6 +391,16 @@ func (t *TFCTrigger) cloneGitRepo(ctx context.Context, mr vcs.MR) (vcs.GitRepo,
...
+// production path), each workspace is enqueued for the workspace worker to
+// process; otherwise we fall back to the inline flow used by tests.
+//
+// Validation failures (allow/deny list, no changes, project missing
+// .tfbuddy.yaml) are returned synchronously so comment-triggered callers can
+// surface them immediately. Per-run failures land in the workspace worker,
+// which posts them to the MR/PR directly.
 func (t *TFCTrigger) TriggerTFCEvents(ctx context.Context) (*TriggeredTFCWorkspaces, error) {
 	ctx, span := otel.Tracer("GitlabHandler").Start(ctx, "TriggerTFCEvents")
 	defer span.End()
```
[the remark this answers is not in the document]
Comment (MeNsaaH) [reply pkg/tfc_trigger/tfc_trigger.go:417]: 💅🏿 , you can update this if you can. This should get the provider name from `t.gl`

File: pkg/hooks/server.go [hunk 2]
```diff
@@ -51,8 +53,26 @@ func StartServer() {
...
+	gh := github.NewGithubClient()
 	tfc := tfc_api.NewTFCClient()
 
+	// Setup VCSClients
+	vcsClients := map[string]vcs.GitClient{
+		"gitlab": gl,
+		"github": gh,
+	}
+
+	// Per-workspace fan-out queue. The MR/PR webhook subscriber publishes one
```
Comment (MeNsaaH) [inline pkg/hooks/server.go]: could we put this this behind a flag? so that it can be disabled optionally. We already support running the old inline  

```go
if TFBUDDY_ENABLE_RUN_QUEUES (or a better name):
  	workspaceStream, err := tfc_trigger.NewWorkspaceStream(js)
	if err != nil {
		log.Fatal().Err(err).Msg("could not configure workspace trigger stream")
	}
	if _, err := tfc_trigger.NewWorkspaceTriggerWorker(workspaceStream, vcsClients, tfc, rs); err != nil {
		log.Fatal().Err(err).Msg("could not start workspace trigger worker")
	}
else:
  workspaceStream = nil
```
something along that line

The implementation down the line suggests that there's an option to disable this e.g when we fallback to inline.

File: pkg/tfc_trigger/tfc_trigger.go [hunk 3]
```diff
@@ -385,69 +431,117 @@ func (t *TFCTrigger) TriggerTFCEvents(ctx context.Context) (*TriggeredTFCWorkspa
...
-				log.Error().Err(err).Msg("could not trigger Run for Workspace")
-				workspaceStatus.Errored = append(workspaceStatus.Errored, &ErroredWorkspace{
-					Name:  cfgWS.Name,
-					Error: fmt.Sprintf("could not trigger Run for Workspace. %v", err),
-				})
-				continue
-			}
-			workspaceStatus.Executed = append(workspaceStatus.Executed, cfgWS.Name)
+		msg := &WorkspaceTriggerMsg{Opts: *t.cfg, Workspace: *cfgWS}
+		if err := t.workspaceStream.Publish(ctx, msg); err != nil {
```
Comment (MeNsaaH) [inline pkg/tfc_trigger/tfc_trigger.go]: nats jetstream has message deduplication enabled that ignores duplicate messages sent within a particular time duration. By default this is 2mins.

If a run on an MR fails quickly and a user retriggers it either by apply or plan via comment or pushes, the new action will not be picked up because of deduplication. We need to add another field to the messages that captures the event id `Idempotency-Key` for gitlab and `X-GitHub-Delivery` for github + workspaceID and have those added to the message as well.

This should prevent refiring on the same event within a short time.

This key should then be set as the dedupe key
```golang
	deliveryId = webhookDeliveryId + workspaceID
	if err := .workstream.Publish(ctx, msg, nats.MsgId(deliveryID)); err != nil {
. . .
```

File: pkg/tfc_trigger/workspace_stream.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+// GetId returns the JetStream message ID, used as a dedup key inside
+// JetStream's deduplication window. A duplicate webhook (or redelivery of
+// the parent MR event) cannot enqueue the same workspace+commit+action twice.
+//
+// The ID is a hex SHA-256 of the canonical fields rather than a delimited
+// string so it can't be accidentally collided by exotic project or workspace
+// names that contain the delimiter.
+func (m *WorkspaceTriggerMsg) GetId(ctx context.Context) string {
```
Comment (MeNsaaH) [inline pkg/tfc_trigger/workspace_stream.go:40]: Use the webhook event header id instead in comment above

File: pkg/tfc_trigger/tfc_trigger.go [hunk 5]
```diff
@@ -192,6 +210,24 @@ func (t *TFCTrigger) GetWorkspace() string {
...
+// traces; falling back to a provider-agnostic name keeps spans honest when
+// VcsProvider is unset (older callers and unit tests).
+func (t *TFCTrigger) tracerName() string {
+	switch t.cfg.VcsProvider {
+	case "github":
+		return "GithubHandler"
+	case "gitlab":
+		return "GitlabHandler"
+	case "bitbucket":
+		return "BitbucketHandler"
```
Comment (MeNsaaH) [inline pkg/tfc_trigger/tfc_trigger.go]: :nit: no bitbucket support, this should be removed as its dead code

File: pkg/tfc_trigger/workspace_worker.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		runstream: w.runstream,
+		cfg:       cfg,
+	}
+
+	mr, err := trigger.gl.GetMergeRequest(ctx, trigger.GetMergeRequestIID(), trigger.GetProjectNameWithNamespace())
+	if err != nil {
+		return fmt.Errorf("could not read MergeRequest data from VCS API: %w", err)
+	}
+
+	repo, err := trigger.cloneGitRepo(ctx, mr)
```
Comment (MeNsaaH) [inline pkg/tfc_trigger/workspace_worker.go:125]: Since each workspace per MR gets a clone,  an MR with loads of workspaces would clone the repo multiple time. A concession that perhaps should be documented, because that would also increase egress traffic/costs.

File: pkg/tfc_trigger/workspace_stream.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+// WorkspaceStream wraps a gongs.GenericStream so callers don't need to import
+// gongs directly to publish or subscribe.
+type WorkspaceStream struct {
+	stream *gongs.GenericStream[WorkspaceTriggerMsg, *WorkspaceTriggerMsg]
+}
+
+// NewWorkspaceStream provisions the JetStream that backs workspace fan-out
+// and returns a publisher/subscriber wrapper.
+func NewWorkspaceStream(js nats.JetStreamContext) (*WorkspaceStream, error) {
+	cfg := &nats.StreamConfig{
```
Comment (MeNsaaH) [inline pkg/tfc_trigger/workspace_stream.go:90]: Other streams have a healthcheck. this should also expose similar healthcheck https://github.com/zapier/tfbuddy/blob/d133120ccb266f7050acf060d5dca8d1da5ff87e/pkg/hooks/server.go#L50

**Review by MeNsaaH — CHANGES_REQUESTED** [review]
overall implementation looks neat. a few edges to tighten

**Comment by MeNsaaH** [comment]
I've just merged #64, you could rebase and fix the conflits. mostly just passing the config.

File: pkg/tfc_trigger/tfc_trigger.go (hunk 3 above)
Comment (chpatinos) [reply pkg/tfc_trigger/tfc_trigger.go]: Fixed: each webhook envelope now threads the upstream delivery ID (X-GitHub-Delivery / Idempotency-Key for GitLab) through TFCTriggerOptions.DeliveryID into the published WorkspaceTriggerMsg.

The dedup key is <DeliveryID>: is a sha256 that includes: project, MR IID, commit, action, vcs, workspace, and org; and it's set as the Nats-Msg-Id, so JetStream keys on it. A fresh push or retrigger comes with a new delivery ID → new key → not dropped. Workspace identity is hashed in too, so fan-out to N workspaces still produces N distinct keys.

File: pkg/tfc_trigger/tfc_trigger.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chpatinos) [reply pkg/tfc_trigger/tfc_trigger.go:417]: Replaced the hard-coded tracer with `t.tracerName()` across all TFCTrigger spans. tracerName() derives the name from t.cfg.VcsProvider so traces are labeled correctly regardless of which stream worker invoked the trigger.

File: pkg/tfc_trigger/workspace_worker.go (hunk 6 above)
Comment (chpatinos) [reply pkg/tfc_trigger/workspace_worker.go:125]: Documented this point in `docs/architecture.md`

File: pkg/tfc_trigger/workspace_stream.go (hunk 7 above)
Comment (chpatinos) [reply pkg/tfc_trigger/workspace_stream.go:90]: Added a `workspace-trigger-stream` liveness check only when workspace fanout is enabled

File: pkg/tfc_trigger/workspace_stream.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+	ctx context.Context
+}
+
+// GetId is the JetStream dedup key. Anchoring on the upstream webhook delivery
+// ID ensures legitimate retriggers (new push/comment/redelivery) produce a
+// fresh key within the dedup window; hashing the rest avoids delimiter
+// collisions when fields contain dashes (GitHub UUIDs, GitLab idempotency keys).
+func (m *WorkspaceTriggerMsg) GetId(ctx context.Context) string {
+	h := sha256.New()
+	enc := json.NewEncoder(h)
```
Comment (MeNsaaH) [inline pkg/tfc_trigger/workspace_stream.go:46]: do we still need all these if we have the deliveryID + TF workspace id? the deliveryID already covers all the gitlab attributes.

Comment (MeNsaaH) [reply pkg/tfc_trigger/workspace_stream.go:46]: seems we don't have tfc workspace id, so I reckon we could do `$DeliverID-$TFCWorkspace.Name-$TFCWorkspace.Org`

Comment (chpatinos) [reply pkg/tfc_trigger/workspace_stream.go:46]: Good call

Updated it, so when DeliveryID is present the key is now just `$DeliveryID/$Workspace.Name/$Workspace.Organization` without hash. The delivery ID already covers all the event-level attributes, so hashing them was redundant noise.

I kept the sha256 fallback for the edge case where DeliveryID might be empty.

Used `/` as separator instead of `-` since TFC workspace names and orgs can contain dashes but never slashes so we avoid the delimiter collision class without needing a hash.

Also added a test guard that asserts empty-DeliveryID messages produce a 64-char sha256 hex, so we catch it if someone accidentally routes them through the plaintext path later.

**Review by MeNsaaH — APPROVED** [review]
